config.MD
# SQM Data Retriever Configuration

The SQM Data Retriever can be customized by editing the file config.php.

All entries in config.php are optional.

## config.php

### Data directory

```$data_directory = <string>;```

The directory relative to sqm.php containing the data files.
		
If using data from a single station, data files should be placed in this directory. If using data from multiple stations, each station should have a subdirectory inside the data directory for its files, e.g. data/sqm1/sqm1_2024-04.dat, data/sqm2/sqm2_2023-08.dat etc.

If a file named .info is placed in the directory with the data files formatted as
```
	Name: My SQM Station
	Latitude: 30.56
	Longitude: -110.23
	Elevation: 80.1
	Timezone: America/Los_Angeles
```
then this information will be used to describe the station.

If no such file exists, the data files will be parsed looking for the same information.  Creating such files is recommended as determining information for a station from data files may be unreliable.

Default is to expect the data files to be in a directory 'data' in the same directory as sqm.php.

### Cache directory

```$cache_directory = <string> | null;```

The directory relative to sqm.php to use for cacheing.  Set this to null to not cache.  If the cache directory is not writeable by the web server user, no cacheing will occur.

Default is the directory 'cache' in the same directory as sqm.php.
		
### Trust or distrust files

```$trust_files = true | false | 'only if not cacheing';```

Whether data files should be trusted to be named including YYYY-MM or YYYY-MM-DD, to have the data in chronological order and to end in newline characters.

This option should not be enabled unless cacheing cannot be enabled and even then should only be enabled if the dataset is so large that not enabling it causes the response time to be too slow.
	
This option should only be used without cacheing and will result in unexpected behavior
if combined with cacheing.

Options are true, false or 'only if not cacheing'.

Default is 'only if not cacheing'.

### Add raw data

```$add_raw_data = true | false | 'only if cacheing';```

Whether to return the raw csv data along with the readings.  If enabled, each reading will have attached to it a 'raw' value which is a key value array keyed by csv column headings.

Options are true, false and 'only if cacheing'.  Default is 'only if cacheing'.

### Add sun moon info

```$add_sun_moon_info = true | false | 'only if cacheing';```

Whether to add sun and moon location information to the data.
		
Options are true, false or 'only if cacheing'; 'only if cacheing' is the preferred choice as this can take time.

Default is 'only if cacheing'.

### Regression analysis

```
$perform_regression_analysis = true | false | 'only if cacheing';
$regression_time_range = <number>;
$regression_averaging_time_range = <number>;
$regression_time_shift = <number>;
```

$perform_regression_analysis controls whether to perform linear regression on the data to attempt to determine cloudiness.  The algorithm used is based on similar logic to that of the algorithm due to Bill Kowalik and adapted by Anthony Tekatch in the Unihedron software's Sun-Moon-MW-Clouds calculation.

Options are true, false or 'only if cacheing'.

Default is 'only if cacheing'.

$regression_time_range specifies the time range in minutes to perform regression over.  $regression_averaging_time_range specifies the time range in minutes to average the regression values over.  $regression_time_shift is the time in minutes after sunset/before sunrise to begin performing regression analysis.

### Filtering

```
$filter_mean_r_squared = <number>;
$filter_sun_elevation = <number>;
$filter_moon_elevation = <number>;
$filter_moon_illumination = <number>;
```

If regression analysis is enabled, $filter_mean_r_squared is the value which any reading with mean $r^2$ above is marked as filtered.  For best nightly readings, both the best overall and the best of those not filtered are returned.

If adding sun and moon information is also enabled, $filter_sun_elevation will filter readings when the sun is above that value in degrees.  $filter_moon_elevation and $filter_moon_illumination in conjunction will filter data based on when the moon is both above that elevation and has illumination greater than the filter value.

In all cases, all readings are returned; the filtering simply means that readings will be marked as filtered and that best nightly readings will include both the best overall and the best of the unfiltered.

### Images

```
$use_images = true | false | 'only if cacheing';
$image_directory = <string>;
$image_directory_url = <string>;
$image_name_format = <string>;
$image_name_prefix_length = <number>;
$image_name_suffix_length = <number>;
$image_time_frame = <number>;
```

If $use_images is set to true (or to 'only if cacheing' and caching is enabled) and $image_directory is set to a directory (default is 'images' in the same directory as sqm.php), the retriever will scan the images folder looking for the image closest in time to the reading and attach the name of that file to the reading.

The directory structure must be images/SQM/YYYY-MM/YYYY-MM-DD/image-file-name where YYYY is the four digit year, MM is the two digit month and DD is the two digit day.  SQM is the name of the corresponding data folder.

$image_directory_url is the url of the images directory (which is simply 'images' in the default case).  $image_name_format is the datetime format of the datetime stamp in the image file name; valid values are those accepted by [PHP Datetime](https://www.php.net/manual/en/datetime.createfromformat.php).

$image_name_prefix_length and $image_name_suffix_length indicate how many characters precede and follow the datetime stamp in the image file names.

For example, the [AllSky software](https://github.com/AllskyTeam/allsky) generates images named e.g. image-20240303175102.jpg so the prefix length is 6, the suffix length is 4 and the format is YmdHis.

$image_time_frame is the time in seconds between when a reading was taken and an image was taken for which the image ought to be considered as connected to the reading; among the images within this time period of a reading, the closest in time will be used.

$use_images may be true, false or 'only if cacheing'.  By default, 'only if cacheing' and the AllSky naming conventions are used.  If $image_directory is not set or the directory does not exist, this is equivalent to $use_images being false.

### Resizing images

```
$resize_images = true | false;
$resized_images_directory = <string>;
$resized_widths = [ <string> => <number>, ... ];
$resized_directory_url = <string>;
```

The retriever can automatically resize images to create e.g. thumbnails.

If $resize_images is set to true and the $resized_images_directory is a writeable directory (default is 'resized_images'), the retriever will resize each image to widths defined by $resized_widths.  $resized_widths is a key value array with entries such as 'thumbnail'=>200 indicating that a directory named 'thumbnail' inside 'resized_images' should be created and contain an image with the same name as each image in 'images' but resized to a width of 200 pixels.

$resized_directory_url is the url of the resized images directory.

Default is false.

### Clear cache on errors

```$clear_cache_on_errors = true | false;```

If set to true, the cache will automatically be cleared when an error in the retriever occurs.  The most likely source of errors in the retriever is cache corruption so this option should be enabled if the application is for public use.

Alternatively, the included scripts for clearing the cache can be used manually if an error occurs and persists.

Default is true.

### Update cache from command line only

```$update_cache_cli_only = true | false;```

If set to true, the cache will only be updated by the command line utility and web access will not trigger reading the files for new data.  Only enable this if you have arranged a cron job or the equivalent to run the command line utility to update the cache.

### Time and memory limits

```
$extended_time = true | false;
$sqm_memory_limit_if_cacheing = '<number>MB';
$sqm_memory_limit_if_not_cacheing = '<number>MB';
```

$extended_time can be set to true to allow the retriever to attempt to override the server's built-in timeout for php scripts.  Especially when resizing images, the initial run can be time consuming
and this option is designed to avoid script timeouts.

$sqm_memory_limit_if_cacheing and $sqm_memory_limit_if_not_cacheing specify how much memory the retriever should request from the server.  If not set, the retriever will work with the memory limit proscribed by the server.

Note that on shared hosting, the server may not allow the retriever to request extended time or extra memory.

### Logging

```$enable_logging = true | false;```

If $enable_logging is true, the retriever will log recoverable errors such as directories not being writeable or data files not being parseable to the server's error log.

config.php
<?php
	/*	config.php
		SQM Data Retriever
		(c) 2024 Darren Creutz
		Licensed under the GNU AFFERO GENERAL PUBLIC LICENSE v3 */
		
	/*	SQM Data Retriever configuration options */

	/*	directory relative to sqm.php containing sqm data files
		
		if using data from a single station, data files should be placed in this directory
		if using data from multiple stations, each station should have a subdirectory for its files
		stations will be identitifed internally by the names of the subdirectories
		
		if a file named .info is placed in the directory with the data files formatted as
			Name: <name>
			Latitude: <latitude>
			Longitude: <longitude>
			Elevation: <elevation>
		then this information will be used to describe the sqm station
		
		if no such file exists, the data files will be parsed looking for the same information
		creating such files is recommended as naming a station from data files is unreliable */
	$data_directory = ".." . DIRECTORY_SEPARATOR . "data";
	
	/*	directory relative to sqm.php to use for cacheing
		
		set to null to not cache
		if cache_directory is not writeable no cacheing will occur */
	$cache_directory = ".." . DIRECTORY_SEPARATOR . "cache";
	
	/*	update cache from command line only
	
		set to true to not load new data into the cache except using the command line utility
		only enable this if you have configured a cron job or equivalent to update the cache */
	$update_cache_cli_only = false;
	
	/*	whether data files should be trusted to be named including YYYY-MM or YYYY-MM-DD
		and trusted to have the data in chronological order and end in newline characters
	
		this option should only be used without cacheing and may result in unexpected behavior
		if combined with cacheing 
		
		options are true, false or 'only if not cacheing' 
		'only if not cacheing' is the preferred choice */
	$trust_files = 'only if not cacheing';
	
	/*	whether to include the raw data from the csv files in the response
	
		options are true, false or 'only if cacheing' */
	$add_raw_data = 'only if cacheing';
	
	/*	whether the backend should add sun and moon location information to the data
		
		options are true, false or 'only if cacheing'
		'only if cacheing' is the preferred choice as this can take time */	
	$add_sun_moon_info = 'only if cacheing';
	
	/*	whether to perform linear regression on the data to attempt to determine cloudiness
	
		options are true, false or 'only if cacheing'
		'only if cacheing' is preferred as this takes significant time */
	$perform_regression_analysis = 'only if cacheing';
	
	/*	time range in minutes for regression analysis
	
		data over this period of time will be used in the linear regression */
	$regression_time_range = 60;
	
	/*	time range in minutes for regression averaging
	
		mean r^2 will be computed over this time period */
	$regression_averaging_time_range = 20;
	
	/*	shift in minutes past sunset/before sunrise to compute mean R^2 values for
		if using twilight type other than astronomical, this is necessary since the sun setting
		causes R^2 values to fluctuate not due to clouds */
	$regression_time_shift = 60;
	
	/*	if regression is performed, data requests will receive both the complete data and
		the data filtered by cloudiness and sun and moon location
		
		filter_mean_r_squared is the r^2 value declared to be cloudy, any reading with a larger
		mean r^2 will be filtered
		
		filter_sun_elevation will filter out data when the sun is at or above that elevation
		
		filter_moon_elevation and filter_moon_illumination determine filtering based on the moon
		readings when the moon is at or above that elevation and has illumination at or above
		the given value will be filtered */
	$filter_mean_r_squared = 0.04;
	$filter_sun_elevation = -12;
	$filter_moon_elevation = -10;
	$filter_moon_illumination = 0.1;
	
	/*	if images are available, such as those produced by https://github.com/AllskyTeam/allsky
		then the backend is able to determine which image is closest to each reading and
		return that information along with the data
		
		use_images determines whether images should be considered
		options are true, false and 'only if cacheing'
		'only if cacheing' is preferred as this takes significant time
		
		image_directory is the directory containing the images
		
		image_directory_url is the url of the image directory
		
		images must have the date and time stamp of when they were taken
		
		image_name_prefix_length is how many characters appear in the file names before
		the datetime stamp
		
		image_name_suffix_length is how many characters appear after the datetime stamp
		
		image_name_format is the datetime format of the datetime stamp
		valid values are those accepted by
		https://www.php.net/manual/en/datetime.createfromformat.php
		
		for example, if using the allsky software, images are named like
			image-20240303175102.jpg
		so prefix length is 6, suffix length is 4
		format is YmdHis corresponding to Year Month Day Hour Minute Second */
	$use_images = 'only if cacheing';
	$image_directory = ".." . DIRECTORY_SEPARATOR . "images";
	$image_directory_url = "images";
	$image_name_format = "YmdHis";
	$image_name_prefix_length = 6;
	$image_name_suffix_length = 4;
	
	/*	the maximum length of time in seconds between when the image was taken and when the
		reading was taken for which the backend will consider pairing them
		if multiple images are within this time span of a reading, the closest will be used */
	$image_time_frame = 1200;
	
	/*	the backend can automatically create resized images such as thumbnails and attach them
		to the data
		
		***NOTE*** the 'gd' php module must be installed for resizing to work
		
		resize_images determines whether resizing is attempted, values are true or false
		
		resized_directory is the directory to save the resized images in
		it must be writeable for resizing to be enabled
		
		resized_widths is an array of name => width pairs governing the resizing
		each image will be resized to each of the widths specified
		the resized image for a given name will be saved with the same name (and relative path)
		as the original image file but in the subdirectory of the resized directory
		which has the given name */
	$resize_images = true;
	$resized_directory = ".." . DIRECTORY_SEPARATOR . "resized_images";
	$resized_directory_url = "resized_images";
	$resized_widths = array(
		'display_image'=>800,
		'thumbnail'=>200
	);
	
	/*	whether to clear the cache if an error occurs
		recommended to be true since cache corruption due to process termination before
		completion is the most likely source of errors
		
		note: the cache is only cleared if the process has the cache lock for obvious reasons */
	$clear_cache_on_errors = true;
	
	/*	whether to override the default script timeout when running
	
		this should be set to true when cacheing since the first pass can take significant time
		if there are large datasets and regression analysis is being performed */
	$extended_time = true;
	
	/*	allowed memory limit
		on some shared hosting, this value may not be modifiable */
	// $sqm_memory_limit_if_cacheing = '256M';
	// $sqm_memory_limit_if_not_cacheing = '512M';
	
	/*	whether to log recoverable errors in the web server error log */
	$logging_enabled = true;
	
	/*	turning on debugging will log errors to the error log
		it will also disable the global error handler meaning that errors in the backend
		may not be properly reported to the front end
		
		this option should not be used except in development */
	$debug = false;
?>
